crates/sui-bcs-json/src/lib.rs

```diff
@@ -0,0 +1,648 @@
+        unsupported => bail!("Requested type {} is currently unsupported.", unsupported),
+    };
+
+    match bcs_bytes {
+        Ok(bcs) => Ok(bcs),
+        Err(e) => bail!("Cannot convert json to bcs, {}", e),
+    }
+}
+
+/// Return 0 for success and 1 for failure
+///
+/// Converts the JSON data into a BCS array.
+/// The result points to the address where the new BCS
+/// array is stored. Don't forget to deallocate the memory
+/// by calling the sui_bcs_json_free_array function.
+#[ffi_export]
+fn sui_bcs_from_json(
+    type_name: char_p::Ref<'_>,
+    json_data: char_p::Ref<'_>,
+    result: *mut *const u8,
+    length: *mut *mut size_t,
+) -> usize {
+    match internal_bcs_from_json(type_name.to_str(), json_data.to_str()) {
+        Ok(output) => {
+            let mut vec = output;
+            vec.shrink_to_fit();
+            assert!(vec.len() == vec.capacity());
+            let ptr = vec.as_mut_ptr();
+            let len = vec.len();
+            std::mem::forget(vec); // prevent deallocation in Rust
+                                   // The array is still there but no Rust object
+                                   // feels responsible. We only have ptr/len now
+                                   // to reach it.
+            unsafe {
+                *result = ptr as *const u8;
+                *length = len as *mut size_t;
+            }
+            return 0;
+        }
+        Err(_) => {
+            return 1;
+        }
+    }
+}
+
+/// Return 0 if the conversion from BCS to JSON is successful, and 1 for
+/// failure.
+///
+/// The result argument will point to the address where the JSON
+/// string is stored. Make sure you release the allocated memory
+/// by calling sui_bcs_json_free_string function!
+#[ffi_export]
+fn sui_bcs_to_json(
+    type_name: char_p::Ref<'_>,
+    bcs_buf_ptr: *mut *mut u8,
+    len: size_t,
+    result: *mut *const safer_ffi::c_char,
+    pretty: bool,
+) -> usize {
+    // we have a pointer to a pointer, thus we first get the address of the pointer
+    // to the bcs array data
+    let ptr = unsafe { std::slice::from_raw_parts(bcs_buf_ptr, 1) };
+    // and then reconstruct the slice
+    let data = unsafe { std::slice::from_raw_parts(ptr[0] as *const u8, len) };
+    match internal_bcs_to_json(type_name.to_str(), &data, pretty) {
+        Ok(res) => {
+            let cstr = std::ffi::CString::new(res).unwrap().into_raw();
+            unsafe { *result = cstr as *const safer_ffi::c_char };
+            return 0;
+        }
+        Err(_) => {
+            return 1;
+        }
+    }
+}
+
+/// Frees a Rust-allocated Vec<u8>.
+#[ffi_export]
+fn sui_bcs_json_free_array(ptr: *mut u8, len: *mut size_t) {
+    let len = len as usize;
+    unsafe {
+        drop(Vec::from_raw_parts(ptr, len, len));
+    }
+}
+
+/// Frees a Rust-allocated string.
+#[ffi_export]
+unsafe fn sui_bcs_json_free_string(pointer: *const safer_ffi::c_char) {
+    drop(std::ffi::CString::from_raw(pointer as *mut i8));
```

Comment:
you don't need to call drop, you just need to build the thing and the compiler will drop it for you